### PR TITLE
Fix animations with multi-bone-skinning

### PR DESCRIPTION
**Pull request by Tarcontar**

### CODE REVIEW

File: src/OpenSage.Game/Graphics/MeshVertex.cs [hunk 1]
```diff
@@ -27,10 +32,15 @@ public struct TexCoords
         {
             new VertexLayoutDescription(
                 new VertexElementDescription("POSITION", VertexElementSemantic.TextureCoordinate, VertexElementFormat.Float3),
+                new VertexElementDescription("POSITION", VertexElementSemantic.TextureCoordinate, VertexElementFormat.Float3),
+                new VertexElementDescription("NORMAL", VertexElementSemantic.TextureCoordinate, VertexElementFormat.Float3),
                 new VertexElementDescription("NORMAL", VertexElementSemantic.TextureCoordinate, VertexElementFormat.Float3),
```
Comment (Tarcontar) [inline src/OpenSage.Game/Graphics/MeshVertex.cs]: Not sure why this is all VertexElementSemantic.TextureCoordinate, shouldnt these be "Position" or sth?

Comment (tgjones) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: It’s correct as it is. Veldrid.SPIRV always maps the semantics as TEXCOORDn.

File: src/OpenSage.Game/Content/ModelLoader.cs [hunk 2]
```diff
@@ -319,17 +319,32 @@ private static FixedFunction.VertexMaterial[] CreateMaterials(W3dMesh w3dMesh)
             {
                 vertices[i] = new MeshVertex.Basic
                 {
-                    Position = w3dMesh.Vertices.Items[i],
-                    Normal = w3dMesh.Normals.Items[i],
+                    Position0 = w3dMesh.Vertices.Items[i],
+                    Position1 = w3dMesh.Vertices2 != null
```
Comment (tgjones) [inline src/OpenSage.Game/Content/ModelLoader.cs]: Is this necessary, or can you fallback to Vector3.Zero? Or put another way, if there are two bone influences, is Positions2 always present?

File: src/OpenSage.Game/Graphics/MeshVertex.cs [hunk 3]
```diff
@@ -9,11 +9,16 @@ public static class MeshVertex
         [StructLayout(LayoutKind.Sequential)]
         public struct Basic
         {
-            public Vector3 Position;
-            public Vector3 Normal;
+            public Vector3 Position0;
```
Comment (tgjones) [inline src/OpenSage.Game/Graphics/MeshVertex.cs]: I think there are now enough skinning-specific attributes that we should have separate structs and code paths for skinning / non-skinning. But that can be done later.

Comment (Qibbi) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: Skinned meshes in at least TW has up to two bones, so wouldn't it make sense to have this a bit more dynamic and create the VertexLayoutDescription according to what is actually there to reduce the array size?

Comment (tgjones) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: I think we should do that, at the same time as we allow different vertex attributes for skinning / non-skinning.

Comment (tgjones) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: The problem will be in keeping the number of shader permutations under control. We don’t have a good way yet of dealing with those.

Comment (Qibbi) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: Doesn't really matter, does it? The shader uses registers which is what the VertexLayoutDescription is for.

Comment (tgjones) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: I assumed there would be warnings or errors if the vertex data doesn’t contain everything expected by the shader. At least, that’s how it works in D3D.

Comment (tgjones) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: We’d need to be able to basically compile multiple entry points per shader, one for each unique set of vertex attributes, and we can’t do that yet. It’s enough work that we will tackle it separately.

Comment (Qibbi) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: I see, TW uses macros and shader arrays for that.

File: src/OpenSage.Game/Graphics/MeshVertex.cs (hunk 1 above)
Comment (Tarcontar) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: ok, but then the other enum values make no sense...

Comment (tgjones) [reply src/OpenSage.Game/Graphics/MeshVertex.cs]: Veldrid can be used without spirv-cross or Veldrid.Spirv. Those enum values can be used, just not by us.

File: src/OpenSage.Game/Content/ModelLoader.cs (hunk 2 above)
Comment (Qibbi) [reply src/OpenSage.Game/Content/ModelLoader.cs]: Yes the second position is always there as it's the same vertex but in the other bone's space.

Comment (Tarcontar) [reply src/OpenSage.Game/Content/ModelLoader.cs]: well how would you test that for every model, need to iterate over every model and over every vertex influence?
Also what is the problem to use the default vertices here?

Comment (Qibbi) [reply src/OpenSage.Game/Content/ModelLoader.cs]: The way they apply the bone influences wouldn't even work without a corresponding set of vertices.

Comment (Qibbi) [reply src/OpenSage.Game/Content/ModelLoader.cs]: `		int index = blendIndices.x * 2;
		WorldPosition = BoneTransformPosition(WorldBones[index], WorldBones[index+1], InSkin.Position0) * InSkin.BlendWeights.x;
		index = blendIndices.y * 2;
		WorldPosition += BoneTransformPosition(WorldBones[index], WorldBones[index+1], InSkin.Position1) * InSkin.BlendWeights.y;`

Comment (Tarcontar) [reply src/OpenSage.Game/Content/ModelLoader.cs]: yea looks like it, wrote some code to test it though, so if the second bone index is not 0, there are always vertices2 and normals2

**Pull request merged**